mm2src/coins/utxo/utxo_common.rs

```diff
@@ -4317,6 +4545,93 @@ where
         .collect()
 }
 
+pub async fn send_dex_fee_with_premium<T>(coin: T, args: SendDexFeeWithPremiumArgs<'_>) -> TransactionResult
+where
+    T: UtxoCommonOps + GetUtxoListOps + SwapOps,
+{
+    let taker_htlc_key_pair = coin.derive_htlc_key_pair(args.swap_unique_data);
+    let total_amount = &args.dex_fee_amount + &args.premium_amount;
+
+    let SwapPaymentOutputsResult {
+        payment_address,
+        outputs,
+    } = try_tx_s!(generate_swap_payment_outputs(
+        &coin,
+        args.time_lock,
+        taker_htlc_key_pair.public_slice(),
+        args.other_pub,
+        args.secret_hash,
+        total_amount,
+        SwapPaymentType::DexFeeWithPremium,
+    ));
+    if let UtxoRpcClientEnum::Native(client) = &coin.as_ref().rpc_client {
+        let addr_string = try_tx_s!(payment_address.display_address());
+        client
+            .import_address(&addr_string, &addr_string, false)
+            .map_err(|e| TransactionErr::Plain(ERRL!("{}", e)))
+            .compat()
+            .await?;
+    }
+    send_outputs_from_my_address(coin, outputs).compat().await
+}
+
+pub async fn validate_dex_fee_with_premium<T>(coin: &T, args: ValidateDexFeeArgs<'_>) -> ValidateDexFeeResult
+where
+    T: UtxoCommonOps + SwapOps,
+{
+    let dex_fee_tx: UtxoTx =
+        deserialize(args.dex_fee_tx).map_to_mm(|e| ValidateDexFeeError::TxDeserialization(e.to_string()))?;
+    if dex_fee_tx.outputs.len() < 2 {
+        return MmError::err(ValidateDexFeeError::TxLacksOfOutputs);
+    }
+
+    let taker_pub =
+        Public::from_slice(args.other_pub).map_to_mm(|e| ValidateDexFeeError::InvalidPubkey(e.to_string()))?;
+    let maker_htlc_key_pair = coin.derive_htlc_key_pair(args.swap_unique_data);
+    let total_expected_amount = &args.dex_fee_amount + &args.premium_amount;
+
+    let expected_amount_sat = sat_from_big_decimal(&total_expected_amount, coin.as_ref().decimals)?;
+
+    let redeem_script = swap_proto_v2_scripts::dex_fee_script(
+        args.time_lock,
+        args.secret_hash,
+        &taker_pub,
+        maker_htlc_key_pair.public(),
+    );
+    let expected_output = TransactionOutput {
+        value: expected_amount_sat,
+        script_pubkey: Builder::build_p2sh(&AddressHashEnum::AddressHash(dhash160(&redeem_script))).into(),
+    };
+
+    if dex_fee_tx.outputs[0] != expected_output {
+        return MmError::err(ValidateDexFeeError::InvalidDestinationOrAmount(format!(
+            "Expected {:?}, got {:?}",
+            expected_output, dex_fee_tx.outputs[0]
+        )));
+    }
```

Comment:
```suggestion
if dex_fee_tx.outputs.first() != Some(&expected_output) {
    return MmError::err(ValidateDexFeeError::InvalidDestinationOrAmount(format!(
        "Expected {:?}, got {:?}",
        expected_output, dex_fee_tx.outputs.first()
    )));
}
```

Reply:
As outputs' length is checked previously, is this really required?
https://github.com/KomodoPlatform/komodo-defi-framework/blob/ebe7e4f36b166c7262ec5b7182a4f16fee22ef46/mm2src/coins/utxo/utxo_common.rs#L4584-L4586

Reply:
> As outputs' length is checked previously, is this really required?

I missed that on the review. At this point it's not required, but still safer way to follow in my opinion(e.g., when a dev makes changes here will likely forget updating the following parts of codes since it's not easy to catch). You can keep it or update it, I leave it to you

Reply:
Yeah, I will update it in the next iteration - we can change the number of outputs in the future, and as you said, forget to update relevant parts of the code. Using `get(index)` will be safer in general.